libpages/server.go

```diff
@@ -29,31 +30,70 @@ import (
 
 // ServerConfig holds configuration parameters for Server.
 type ServerConfig struct {
-	DomainWhitelist  []string
+	// If DomainWhitelist is non-nil and non-empty, only domains in the
+	// whitelist are served and others are blocked.
+	DomainWhitelist []string
+	// If DomainBlacklist is non-nil and non-empty, domains in the blacklist
+	// and all subdomains under them are blocked. When a domain is present in
+	// both blacklist and whitelist, the domain is blocked.
+	DomainBlacklist  []string
 	UseStaging       bool
 	Logger           *zap.Logger
 	UseDiskCertCache bool
 	StatsReporter    StatsReporter
 
-	domainWhitelistOnce sync.Once
-	domainWhitelist     map[string]bool
+	domainListsOnce sync.Once
+	domainWhitelist map[string]bool
+	domainBlacklist []string
 }
 
-func (c *ServerConfig) allowedByDomainWhiteList(domain string) bool {
-	c.domainWhitelistOnce.Do(func() {
+// ErrDomainBlockedInBlacklist is returned when the server is configured
+// with a domain whitelist, and we receive a HTTP request that was sent to a
+// domain that's not in the whitelist.
+type ErrDomainBlockedInBlacklist struct{}
+
+// Error implements the error interface.
+func (ErrDomainBlockedInBlacklist) Error() string {
+	return "a blacklist is configured and the given domain is in the list"
+}
+
+// ErrDomainNotAllowedInWhitelist is returned when the server is configured
+// with a domain whitelist, and we receive a HTTP request that was sent to a
+// domain that's not in the whitelist.
+type ErrDomainNotAllowedInWhitelist struct{}
+
+// Error implements the error interface.
+func (ErrDomainNotAllowedInWhitelist) Error() string {
+	return "a whitelist is configured and the given domain is not in the list"
+}
+
+func (c *ServerConfig) checkDomainLists(domain string) error {
+	c.domainListsOnce.Do(func() {
 		if len(c.DomainWhitelist) > 0 {
 			c.domainWhitelist = make(map[string]bool, len(c.DomainWhitelist))
 			for _, d := range c.DomainWhitelist {
-				c.domainWhitelist[d] = true
+				c.domainWhitelist[strings.ToLower(strings.TrimSpace(d))] = true
+			}
+		}
+		if len(c.DomainBlacklist) > 0 {
+			c.domainBlacklist = make([]string, len(c.DomainBlacklist))
+			for i, d := range c.DomainBlacklist {
+				c.domainBlacklist[i] = strings.ToLower(strings.TrimSpace(d))
 			}
 		}
 	})
-	if c.domainWhitelist != nil {
-		return c.domainWhitelist[domain]
+
+	for _, blocked := range c.domainBlacklist {
+		if strings.HasSuffix(domain, blocked) {
+			return ErrDomainBlockedInBlacklist{}
+		}
+	}
+	if c.domainWhitelist != nil && !c.domainWhitelist[domain] {
```

Comment:
Maybe change `c.domainWhitelist != nil` to `len(c.domainWhitelist) > 0`?  At first I thought it was an unnecessary nil check, but then I realized you were trying to check if the whitelist was configured at all. My suggestion might be a little clearer for that, unless you want to support the case where an empty whitelist blocks EVERYTHING!  (Which is fine too, but might need a comment here.)